go-mode.el

```diff
@@ -1562,6 +1568,185 @@ for."
       (if (not (eq cur-buffer (current-buffer)))
           (display-buffer (current-buffer) `(,go-coverage-display-buffer-func))))))
 
+(defun go-goto-function ()
+  "Go to the function defintion above point.
+
+If we are on a docstring, follow the docstring down.
+If no function is found, assume that we are at the top of a file
+and search forward instead.
+
+If we are inside an anonymous function, go to that.
+If we are inside a function that has an anonymous function inside of it and we
+are below that anonymous function, go to the root function."
+  (interactive)
+  (let ((p (point)))
+    (cond
+     ((save-excursion
+        (beginning-of-line)
+        (looking-at "^//"))
+      ;; In case we are looking at the docstring, move on forward until we are
+      ;; not anymore
+      (beginning-of-line)
+      (while (looking-at "^//")
+        (forward-line 1))
+      ;; If we are still not looking at a function, retry by calling self again.
+      (when (not (looking-at "^func"))
+        (go-goto-function)))
+
+     ((not (looking-at "^func"))
+      ;; If we are not looking at the beginning of a function line, do a regexp
+      ;; search backwards
+      (re-search-backward "\\<func\\>" nil t)
+
+      ;; If nothing is found, assume that we are at the top of the file and
+      ;; should search forward instead.
+      (when (not (looking-at "func"))
+        (re-search-forward "\\<func\\>" nil t)
+        (forward-word -1))
+
+      ;; If we have landed at an anonymous function, it is possible that we
+      ;; were not inside it but below it. If we were not inside it, we should
+      ;; go to the containing function.
+      (while (go--in-function-p p)
```

Comment:
It feels to me like `go--in-function-p` returns whether `compare-point` is *NOT* in the function that point is at. Consider this piece of code:

```
package main

func main() {
	_ = func() {
		//
	}
	_ = 1
}
```

With `compare-point` on `1` and point looking at the anonymous `func`, `go--in-function-p` returns t. With `point` looking at the named func, it returns nil. Shouldn't the opposite be the case? `compare-point` isn't in the anonymous function, but it is in the named function. 

That's why the while loop feels confusing to me, too. If `go--in-function-p` worked as I expected, the while loop would have to check for `(not (go--in-function-p))`.

What am I missing?